### PR TITLE
8317466: Enable interpreter oopMapCache for concurrent GCs

### PR DESCRIPTION
**Pull request by zhengyu123**
Interpreter oop maps are computed lazily during GC root scan and they are expensive to compute.

GCs uses a small hash table per instance class to cache computed oop maps during STW root scan, but not for concurrent root scan.

This patch is intended to enable `OopMapCache` for concurrent GCs.

Test:
  tier1 and tier2 fastdebug and release on MacOSX, Linux 86_84 and Linux 86_32.

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8317466](https://bugs.openjdk.org/browse/JDK-8317466): Enable interpreter oopMapCache for concurrent GCs (**Enhancement** - P4)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16074/head:pull/16074` \
`$ git checkout pull/16074`

Update a local copy of the PR: \
`$ git checkout pull/16074` \
`$ git pull https://git.openjdk.org/jdk.git pull/16074/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16074`

View PR using the GUI difftool: \
`$ git pr show -t 16074`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16074.diff">https://git.openjdk.org/jdk/pull/16074.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16074#issuecomment-1750696208)

### CODE REVIEW

File: src/hotspot/share/interpreter/oopMapCache.cpp [hunk 1]
```diff
@@ -500,8 +501,10 @@ void OopMapCache::flush_obsolete_entries() {
           ("flush: %s(%s): cached entry @%d",
            entry->method()->name()->as_C_string(), entry->method()->signature()->as_C_string(), i);
       }
-      _array[i] = nullptr;
-      OopMapCacheEntry::deallocate(entry);
+
+      if (Atomic::cmpxchg(&_array[i], entry, (OopMapCacheEntry*)nullptr, memory_order_relaxed) == entry) {
+        enqueue_for_cleanup(entry);
```
Comment (leelamv) [inline src/hotspot/share/interpreter/oopMapCache.cpp:506]: OopMapCache::flush_obsolete_entries() is called from VM_RedefineClasses::doit(). Enqueuing  OopMapCacheEntries for cleanup means that we would need call OopMapCache::cleanup_old_entries() in VM_RedefineClasses::doit_epilogue()

Comment (zhengyu123) [reply src/hotspot/share/interpreter/oopMapCache.cpp:506]: Good catch. Fixed.

File: src/hotspot/share/gc/shenandoah/shenandoahVMOperations.cpp [hunk 2]
```diff
@@ -58,6 +58,10 @@ void VM_ShenandoahFinalMarkStartEvac::doit() {
   _gc->entry_final_mark();
 }
 
+void VM_ShenandoahFinalMarkStartEvac::doit_epilogue() {
+  OopMapCache::cleanup_old_entries();
+}
+
```
Comment (leelamv) [inline src/hotspot/share/gc/shenandoah/shenandoahVMOperations.cpp:64]: Do you think, VM_ShenandoahFinalMarkStartEvac walks the stack roots.  If yes, i recommend adding OopMapCache::cleanup_old_entries() in  VM_ShenandoahOperation::doit_epilogue().  And this would made the change simple and also revert the change in this [PR](https://github.com/openjdk/jdk/pull/15921)

**Review by leelamv — CHANGES_REQUESTED** [review]

File: src/hotspot/share/gc/shenandoah/shenandoahVMOperations.cpp (hunk 2 above)
Comment (zhengyu123) [reply src/hotspot/share/gc/shenandoah/shenandoahVMOperations.cpp:64]: No, `VM_ShenandoahFinalMarkStartEvac ` does not walk the stack roots, it signals the end of mark phase.

File: src/hotspot/share/oops/method.cpp [hunk 3]
```diff
@@ -308,14 +308,7 @@ int Method::fast_exception_handler_bci_for(const methodHandle& mh, Klass* ex_kla
...
   methodHandle h_this(Thread::current(), this);
-  // Only GC uses the OopMapCache during thread stack root scanning
-  // any other uses generate an oopmap but do not save it in the cache.
-  if (Universe::heap()->is_gc_active()) {
-    method_holder()->mask_for(h_this, bci, mask);
-  } else {
-    OopMapCache::compute_one_oop_map(h_this, bci, mask);
-  }
-  return;
+  method_holder()->mask_for(h_this, bci, mask);
```
[the remark this answers is not in the document]
Comment (zhengyu123) [reply src/hotspot/share/oops/method.cpp:311]: The `oopMapCache` now is fully concurrent, it can be used/modified by Java threads.

File: src/hotspot/share/gc/shenandoah/shenandoahVMOperations.cpp (hunk 2 above)
Comment (leelamv) [reply src/hotspot/share/gc/shenandoah/shenandoahVMOperations.cpp:64]: Note we don't need to call   OopMapCache::cleanup_old_entries() after this [PR](https://github.com/openjdk/jdk/pull/15921).  VM_ShenandoahFinalMarkStartEvac is derived from VM_ShenandoahReferenceOperation. VM_ShenandoahReferenceOperation::doit_epilogue calls OopMapCache::cleanup_old_entries();

File: src/hotspot/share/oops/method.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (leelamv) [reply src/hotspot/share/oops/method.cpp:311]: Which thread takes up responsibility of cleaning up old entries accumulated by java threads.  Allowing for java threads definitely increases the scope (than it is mentioned in the bug ).

**Review by leelamv — COMMENTED** [review]
Before of this change,  VM_Operations which enabled "is_gc_active" and walks stack to collect roots had a call to OopMapCache::cleanup_old_entries() in its corresponding doit_epilogue().  This is to free the old entries which are collected during safepoint. 


Now scope of the change isn't clear.

We seem to extend them to concurrent GCs during their concurrent phases (Not just safepoints of concurrent GCs).  Calling OopMapCache::cleanup_old_entries() in doit_epilogue() would effectively cleanup old entries accumulated during concurrent phase of the GC and also during safepoint.  

But change also allows java threads to accumulate old entries. When/Who calls cleanup_old_entries() in this case ? These needs to wait until future GC which does cleanup in doit_epilogue(). However, at least theoretically, we can have large time windows without GCs. Old entries accumulated by java threads can be seen as a leak (until next GC)


A separate thought  to make change simpler. Can cleanup_old_entries become a [list of Cleanup tasks](https://github.com/openjdk/jdk/blob/8d9a4b43f4fff30fd217dab2c224e641cb913c18/src/hotspot/share/runtime/safepoint.hpp#L72) VM thread does during cleanup phase.

**Comment by zhengyu123** [comment]
> Before of this change, VM_Operations which enabled "is_gc_active" and walks stack to collect roots had a call to OopMapCache::cleanup_old_entries() in its corresponding doit_epilogue(). This is to free the old entries which are collected during safepoint.
> 
> Now scope of the change isn't clear.
> 
> We seem to extend them to concurrent GCs during their concurrent phases (Not just safepoints of concurrent GCs). Calling OopMapCache::cleanup_old_entries() in doit_epilogue() would effectively cleanup old entries accumulated during concurrent phase of the GC and also during safepoint.
> 
> But change also allows java threads to accumulate old entries. When/Who calls cleanup_old_entries() in this case ? These needs to wait until future GC which does cleanup in doit_epilogue(). However, at least theoretically, we can have large time windows without GCs. Old entries accumulated by java threads can be seen as a leak (until next GC)
> 
Yes, it could accumulate a few. But from what I saw, there are not many. They are cleanup by:
- Serial, Parallel and G1: any `VM_GC_Operation`s
- ZGC: any `VM_ZOperation`s
- Shenandoah: I should have added `OopMapCache::cleanup_old_entries()` in `VM_ShenandoahOperation`, where should help jdk11u backport.

> A separate thought to make change simpler. Can cleanup_old_entries become a [list of Cleanup tasks](https://github.com/openjdk/jdk/blob/8d9a4b43f4fff30fd217dab2c224e641cb913c18/src/hotspot/share/runtime/safepoint.hpp#L72) VM thread does during cleanup phase.

No. There is no point to add safepoint latency when it can not done outside of safepoint.
